### PR TITLE
[Refactor & Bugfix] Improvements in automatic suspension and deletion of tenants

### PR DESCRIPTION
**Pull request by Martouta**
Closes [THREESCALE-2132](https://issues.jboss.org/browse/THREESCALE-2132)
It blocks https://github.com/3scale/porta/pull/745 because if we merge that one first, it means that we will destroy accounts that shouldn't be destroyed.
The config edited in the _deploy_ repo: https://github.com/3scale/deploy/pull/554 (tested in preview)

### Tasks
- [X] Make the plans that make an account not able to be automatically destroyed customisable instead of hardcoding 'enterprise' by default in the code. Avoid the term `enterprise` in the code and make it configurable. Requested in https://github.com/3scale/porta/pull/555#discussion_r251484561
- [X] Move the complex conditions to auto-suspend and auto-schedule_for_deletion away from the workers.
- [X] **Bugfix**. These application plans whose tenants shouldn't be deleted should be when the tenant is the buyer of that cinstance, not when the tenant is the provider!
- [X] We don't validate that the config from the sysadmin is correct anymore. If there is a config, it is considered correct and feature enabled. If there is no config, feature disabled.
- [X] A rake task to change from `scheduled_for_deletion` to `suspended` to accounts that have already being marked as `scheduled_for_deletion` but shouldn't be 😄 Once this PR is merged and deployed in `production`, I will have to run: `RAILS_ENV=production bundle exec rake multitenant:tenants:suspend_forbidden_plans_scheduled_for_deletion`

### CODE REVIEW

File: app/models/account.rb [hunk 1]
```diff
@@ -277,6 +282,24 @@ def config
   scope :created_before, ->(date) { where(['created_at <= ?', date]) }
   scope :created_after,  ->(date) { where(['created_at >= ?', date]) }
 
+  def self.should_be_automatically_suspended
```
Comment (macejmic) [inline app/models/account.rb]: consider moving this code to `app/queries`.

File: app/lib/features/account_deletion_config.rb [hunk 2]
```diff
@@ -29,14 +34,14 @@ def initialize(config = {})
     private
 
     def format_config(config)
-      config = (config || {}).each_with_object({}) do |(key, value), collection|
-        collection[key.to_sym] = value if value.is_a?(Integer)
+      config = (config || {}).symbolize_keys.slice(*EXPECTED_VALUES.keys).each_with_object({}) do |(key, value), collection|
```
Comment (macejmic) [inline app/lib/features/account_deletion_config.rb]: what does `slice` with has values? can you explain why we need these `EXPECTED_VALUES` ?

File: app/lib/features/account_deletion_config.rb [hunk 3]
```diff
@@ -29,14 +34,14 @@ def initialize(config = {})
...
+        collection[key] = value if value.is_a?(EXPECTED_VALUES[key])
```
Comment (guicassolato) [inline app/lib/features/account_deletion_config.rb]: I still don't see the relevance of this type-checking. This is anti-pattern in Ruby.

File: app/lib/features/account_deletion_config.rb (hunk 2 above)
Comment (Martouta) [reply app/lib/features/account_deletion_config.rb]: @macejmic @guicassolato 
When this feature was introduced, there was a big debate about what to do if someone introduces wrong values in the config. It all started in https://github.com/3scale/porta/pull/574#discussion_r253816442
And later we kept talking between these options if the user introduces wrong config:
- Not booting at all
- Booting without this feature enabled and just showing a warning informing of that
- Let it raise errors (for example if it is not an integer and it does `days.ago` over it) when it is going to be used.

And back then we decided to do the 2nd one. To do the 2nd one, this is the way to do it 🤷‍♀️ If you have a better idea, please elaborate about it 😝

File: app/lib/features/account_deletion_config.rb (hunk 3 above)
Comment (Martouta) [reply app/lib/features/account_deletion_config.rb]: Please let's keep this conversation in one single place https://github.com/3scale/porta/pull/583#discussion_r272113958

File: app/lib/features/account_deletion_config.rb (hunk 2 above)
Comment (guicassolato) [reply app/lib/features/account_deletion_config.rb]: The best way to do it IMO is "If the configs are there, they should be considered valid."

File: app/models/account.rb (hunk 1 above)
Comment (Martouta) [reply app/models/account.rb]: That sounds like a good idea 👍 I will check about it 👍

File: app/lib/features/account_deletion_config.rb (hunk 2 above)
Comment (Martouta) [reply app/lib/features/account_deletion_config.rb]: Ok, spoken in the team, it is decided: "If the configs are there, they should be considered valid."

File: app/lib/features/account_deletion_config.rb [hunk 4]
```diff
@@ -19,31 +15,18 @@ def self.config
...
-      end
-      @config = ActiveSupport::OrderedOptions.new.merge(config)
-    end
-
-    def check_validity(initial_config)
-      @valid = (EXPECTED_VALUES.keys.length == config.length)
-      return if valid || initial_config.blank?
-      Rails.logger.warn '[WARNING] Can\'t enable "automatic inactive tenant account deletion". Please revise your config"'
+    def enabled?
+      !config.length.zero?
```
Comment (guicassolato) [inline app/lib/features/account_deletion_config.rb]: `config.length.positive?`

Comment (hallelujah) [reply app/lib/features/account_deletion_config.rb]: config.any?

**Comment by Martouta** [comment]
## Decision on features with rolling updates or only rolling updates

Until the _January 27th, 2019_, when we wanted to enable/disable a feature for some specific accounts, we used rolling updates. Before that, when we wanted to make a feature enabled/disabled for SaaS vs on-prem, we used a conditional such as `if ThreeScale.master_on_premises?` but we decided to start getting rid of those and start using rolling updates for those cases, as @hallelujah said:

> Well think about we have 3 products in one
> 
> - SaaS
> - On-prem
> - OSS
> Agree, priorities are given to the 2 tops
> but we should not neglect OSS
> For example if ThreeScale.on_premises is something we should remove

We also wanted to make this feature (Account Deletion) configurable instead of having the values in the code.

The `February 5th, 2019`, I first did the code as enabled/disabled with a rolling update but we spoke about the possibility of creating a new kind of configuration called `features`. I quote how @hallelujah started this conversation:

> you could put it in the rolling_updates btw but it applies to master so not really 
> well rolling_updates is: provider_can_use?

So we all spoke about it and back then decided to have `features` for global features of the system instead of 'by provider'. If the config is given, it is enabled and it uses that config, if it isn't, it is disabled.

So I changed this to make it with the `features`, and I also made the [segment deletion PR](https://github.com/3scale/porta/pull/717) as feature because this was the logic that we decided to follow from now on.

So @hallelujah and @guicassolato suggested today to go back to only rolling updates and get rid of the features thing. So one of the opinions that came up was:

> IDK… in that direction, maybe we should get rid of features entirely
> because rolling updates are already exposed as a configmap in on-prem…
> meaning that it’s not only something we use to roll out new features to saas customers
> it’s already a product feature itself


Imo personally the differentiation of features vs rolling updates makes sense, but I am ok with making it all back to rolling updates if it is what we decide to do from now. But I will wait for @hallelujah confirmation here before doing so 😄

File: app/models/cinstance.rb [hunk 5]
```diff
@@ -182,6 +182,8 @@ def self.by_service(service)
   scope :by_active_since, lambda {|date| where('first_daily_traffic_at >= ?', date) }
   scope :by_inactive_since, lambda {|date| where('first_daily_traffic_at <= ?', date) }
 
+  scope :with_application_plans_with_system_names, ->(system_names) { joins(:plan).where.has { plan.system_name.in system_names } }
```
Comment (guicassolato) [inline app/models/cinstance.rb]: How about naming the scope `by_plan_system_names`?

File: app/models/account.rb [hunk 6]
```diff
@@ -110,7 +110,12 @@ def self.columns
   before_destroy :destroy_features
 
   scope :free, ->(free_date) { where.has { not_exists Contract.have_paid_on(free_date).by_account(BabySqueel[:accounts].id).select(:id) } }
-  scope :not_enterprise, -> { where.has { not_exists Service.of_account(BabySqueel[:accounts].id).with_enterprise_application_plans.select(:id) } }
+
+  scope :without_bought_application_plans_with_system_names, lambda { |system_names|
```
Comment (guicassolato) [inline app/models/account.rb]: Why can't I think of a better name to simplify this scope? Any other idea?

File: app/lib/features/account_deletion_config.rb [hunk 7]
```diff
@@ -14,31 +15,18 @@ def self.config
...
-      @config = ActiveSupport::OrderedOptions.new.merge(config)
-    end
+    attr_reader :config
 
-    def check_validity(initial_config)
-      @valid = %i[account_suspension account_inactivity contract_unpaid_time].all? { |key| config.key?(key) }
-      return if valid || initial_config.blank?
-      Rails.logger.warn '[WARNING] Can\'t enable "automatic inactive tenant account deletion". Please revise your config"'
+    def enabled?
+      config.any?
```
Comment (hallelujah) [inline app/lib/features/account_deletion_config.rb]: Hmm I see. 
Well in all our other features, the file always has an option `enabled` https://github.com/3scale/porta/blob/master/config/examples/redhat_customer_portal.yml#L2

I think we should follow that

So this is just delegating to `config.enabled`

File: app/models/account.rb (hunk 6 above)
Comment (Martouta) [reply app/models/account.rb]: my only idea is changing `bought_application_plans` for `cinstances` or `applications` so it is a bit shorter 😂

File: lib/tasks/multitenant/tenants.rake [hunk 8]
```diff
@@ -9,5 +9,15 @@ namespace :multitenant do
         end
       end
     end
+
+    task suspend_forbidden_plans_scheduled_for_deletion: :environment do
+      return unless Features::AccountDeletionConfig.enabled?
```
Comment (guicassolato) [inline lib/tasks/multitenant/tenants.rake]: For the rake task, better adding an output message when it's disabled, like "Account deletion is disabled. Nothing to do.".

**Comment by Martouta** [comment]
### Test in preview
I tested the rake task because the point of testing this in preview in particular is seeing the integration between this PR and the deploy PR.

To do that, first I check the result of:
```ruby
Account.tenants.scheduled_for_deletion.where.has do
  exists Cinstance.by_account(BabySqueel[:accounts].id).by_plan_system_name(Features::AccountDeletionConfig.config[:disabled_for_app_plans]).select(:id)
end.count
```
That is the condition inside the rake task.
The result is `381`.
Then I run the rake task and it works:
image
Finally, I do again the query I tried above and I see `0` 💃 .

**Pull request merged**